Engine/csound_type_system.c

```diff
@@ -314,59 +412,180 @@ static int32_t copy_var_no_op(CSOUND *csound, void *p) {
   return OK;
 }
 
-#include "csound_standard_types.h"
-#include "arrays.h"
+/**
+ * Helper function to normalize type names for comparison.
+ * Converts both ":TypeName;" and "TypeName" to the same format for comparison.
+ * Returns a newly allocated string that must be freed by the caller.
+ */
+static char* normalize_type_name_for_comparison(CSOUND *csound, const char* typeName) {
+  if (!typeName) return NULL;
+
+  // If it's already in internal format (:TypeName;), extract the core name
+  if (typeName[0] == ':' && typeName[strlen(typeName) - 1] == ';') {
+    size_t len = strlen(typeName) - 2; // Remove : and ;
+    char* result = csound->Malloc(csound, len + 1);
+    memcpy(result, typeName + 1, len);
+    result[len] = '\0';
+    return result;
+  }
+
+  // Otherwise, return a copy of the original name
+  return cs_strdup(csound, typeName);
+}
+
+/**
+ * Check if two type names represent the same type, handling both
+ * internal format (:TypeName;) and external format (TypeName).
+ */
+static int types_are_equivalent(CSOUND *csound, const char* type1, const char* type2) {
```

Comment:
Little concerned about the mixing of internal and external formats and the need to compare. Just noting that here that we might want to revisit in the future.

Reply:
yeh, it's not pretty. I think we could keep internal internal and external external. I could open follow up ticket.